src/Features/Core/Portable/PullMemberUp/MembersPuller.cs

```diff
@@ -0,0 +1,414 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.  
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.PullMemberUp;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+using EditorMap = System.Collections.Immutable.ImmutableDictionary<Microsoft.CodeAnalysis.SyntaxTree, Microsoft.CodeAnalysis.Editing.DocumentEditor>;
+using SyntaxMap = System.Collections.Immutable.ImmutableDictionary<Microsoft.CodeAnalysis.ISymbol, Microsoft.CodeAnalysis.SyntaxNode[]>;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp
+{
+    internal class MembersPuller
+    {
+        internal static readonly MembersPuller Instance = new MembersPuller();
+
+        private MembersPuller()
+        {
+        }
+
+        /// <summary>
+        /// Return the CodeAction to pull selectedMember up to destinationType. If the pulling will cause error,
+        /// it will return null.
+        /// </summary>
+        internal CodeAction TryComputeCodeAction(
+            Document document,
+            ISymbol selectedMember,
+            INamedTypeSymbol destinationType)
+        {
+            var result = PullMembersUpAnalysisBuilder.BuildAnalysisResult(destinationType, ImmutableArray.Create((member: selectedMember, makeAbstract: false)));
+            if (result.PullUpOperationCausesError ||
+                IsSelectedMemberDeclarationAlreadyInDestination(destinationType, selectedMember))
+            {
+                return default;
+            }
+
+            return new SolutionChangeAction(
+                string.Format(FeaturesResources.Add_to_0, result.Destination),
+                cancellationToken => PullMembersUpAsync(result, document, cancellationToken));
+        }
+
+        /// <summary>
+        /// Return the changed solution if all changes in result are applied.
+        /// </summary>
+        /// <param name="result">Contains the members to pull up and all the fix operations</param>>
+        internal async Task<Solution> PullMembersUpAsync(
+            PullMembersUpAnalysisResult result,
+            Document contextDocument,
+            CancellationToken cancellationToken)
```

Comment:
meta comment about parameters.  In general (outside of cancellation tokesn) we try to preserve parameter order.  So, the caller of this is a method that takes a document first.  Taht callign method computes teh result, and passes that in.  As such, you can think of 'result' having a narrower lifetime/scope than 'document'.  So this should be:

`PullMembersUpAsync(Document, PullMembersUpResult, CancellationToken)`

(this applies to a lot of the functions in this PR).

Reply:
thanks.  this can be resolved.